### ISSUE TITLE
Support explicit transactions

### ISSUE THREAD

**Issue by rommsen** [Zaid-Ajaj/Npgsql.FSharp#30]
It would be great if we could add support for explicit custom transactions (http://www.npgsql.org/doc/transactions.html) to issue multiple nonqueries within one transaction.  As far as I can see all executeNonQuery functions always just use `List.head SqlProps.queries`.

I am not sure about the API for this feature. It could be `let executeNonQueryMany (props: SqlProps) : int list =`

What do you think?

**Comment by Zaid-Ajaj** [comment]
Hello Roman,

Indeed, this signature of `executeNonQueryMany` does look consistent with the rest of the naming conventions but this assumes that every query that is supplied in `Sql.queryMany` will have the same parameters because `Sql.parameters` does not differentiate between different queries, at least with the way the library is written as of now.

There are a couple of scenario's to take into account with transactions:

### 1- One query (same parameter names) with different values:
Query:
```sql
INSERT INTO "todos" VALUES (@description, @done)
``` 
 Paramters: 
`@description` and `@done` 
Values:
("desc 1", true)
("desc 2", false)
("desc 3", true)

### 2 - Multiple queries (different parameters) with different values
Queries:
```sql
INSERT INTO ... VALUES (@value1, @value2)
UPDATE ... SET X = @value3
```
Paramters:
First query: `@value1` and `@value2`
Second Query: `@value3`

Values:
`you get the idea`

### Proposition
Only support scenario 1 (one query, same parameters, multiple paramter values). This is what #16 suggested too. 

Adding something like `Sql.paramtersMany` to be used in combincation with `Sql.query` and `Sql.executeNonQueryMany`:
```fs
let parameters = 
    [ 1 .. 5 ] 
    |> List.map (fun value -> [ "@number", SqlValue.Int value ]) 

connectionString
|> Sql.connect
|> Sql.query "INSERT INTO ... VALUES (@number)"
|> Sql.paramtersMany parameters  
|> Sql.executeNonQueryMany // uses a transaction, returns int list
```

If someone finds himself/herself needing the second scenario or something even more customized, then it might be easier to just fallback to the raw `Npgsql` API's instead because this library is intended to make the common scenario's as simple as possible. That is of course unless someone has a better and simple way of handling more generic transactions.

What do you think?

**Comment by rommsen** [comment]
Hey Zaid, 

for me both propositions are very useful. I have a use case for both right now (for my EventSourcing video series)

For 1: Make one or many Upserts in a given readmodel
For 2: Within the update I also need to update another table with some meta information and I need to make sure both queries are within one transaction.

While I would like to have both I see no easy way to make proposition 2 work atm so you might be right falling back to Npgsql.

**Comment by rommsen** [comment]
Thinking of it. What do you think of the following API for proposition 2?

```
connectionString
|> Sql.connect
|> Sql.queries
        [ 
            "INSERT INTO ... VALUES (@number)", [ "@number", SqlValue.Int value ]
            "UPDATE ... SET meta = @meta AND time = @time",  [ "@meta", SqlValue.string value ; " @time, SqlValue.... ]
|> Sql.executeNonQueriesTransactional // uses a transaction, returns int list

```

**Comment by Zaid-Ajaj** [comment]
I really like this API, this generalizes the first proposition because you can simply duplicate the query with different parameters (internally I can call `List.groupBy fst`)
```fs
connectionString
|> Sql.connect
|> Sql.executeTransaction // SqlProps -> int list
      [
             "INSERT INTO ... VALUES (@number)", [ "@number", SqlValue.Int 1 ]
             "INSERT INTO ... VALUES (@number)", [ "@number", SqlValue.Int 2 ]
             "INSERT INTO ... VALUES (@number)", [ "@number", SqlValue.Int 3 ]
             "UPDATE ... SET meta = @meta",  [ "@meta", SqlValue.String value ]
      ]
```
Or without duplicates:
```fs
connectionString
|> Sql.connect
|> Sql.executeTransaction // SqlProps -> int list
      [
         "INSERT INTO ... VALUES (@number)", [
             [ "@number", SqlValue.Int 1 ]   
             [ "@number", SqlValue.Int 2 ]
             [ "@number", SqlValue.Int 3 ]
         ]

         "UPDATE ... SET meta = @meta",  [
            [ "@meta", SqlValue.String value ]
         ] 
  ]
```
Then there could be `Sql.executeTransactionSafe`,  `Sql.executeTransactionSafeAsync` etc.

**Comment by rommsen** [comment]
The last one is neat. I think that is a simple enough API to keep the library clean and it would make it tremendously more useful.


About queries that require intermediate results: you are right but maybe this would really be something to fall back to Npgsql?

**Comment by Zaid-Ajaj** [comment]
On a side-note that might be relevant to your Event-Sourcing use-case and it's highly likely you already have looked into it but just in case you missed it:  [CosmoStore](https://github.com/Dzoukr/CosmoStore) has recently published a PostgreSQL backing API built ontop of [Marten](https://github.com/JasperFx/marten), see [CosmoStore.Marten](https://www.nuget.org/packages/CosmoStore.Marten/) package

> maybe this would really be something to fall back to Npgsql?

Yes, anything that complex can be done with multiple queries using this library or a multiple raw queries directly at the `Npgsql` level

**Comment by rommsen** [comment]
Yeah I have checked it out. But I would really like to stick to the DIY approach in the videos and only use minimal libraries, nothing that solves too many problems at once. That is why I would really love to use this one here.

**Comment by Zaid-Ajaj** [comment]
Done! Added `Sql.executeTransaction` and friends in version 1.7.0 (Async, Safe, SafeAsync) and will be available soon. Implemented it like this:
```fs
let executeTransaction queries (props: SqlProps)  = 
    if List.isEmpty queries  
    then [ ]
    else 
    use connection = newConnection props
    connection.Open()
    use transaction = connection.BeginTransaction()
    let affectedRowsByQuery = ResizeArray<int>()
    for (query, parameterSets) in queries do
        for parameterSet in parameterSets do
            use command = new NpgsqlCommand(query, connection, transaction)
            populateRow command parameterSet
            let affectedRows = command.ExecuteNonQuery() 
            affectedRowsByQuery.Add affectedRows
    List.ofSeq affectedRowsByQuery
```
Try it out and please let me know if any problem comes up!

**Comment by Zaid-Ajaj** [comment]
Wait, looks like I forgot to `Commit()` my bad 😅

**Comment by Zaid-Ajaj** [comment]
Alright, please try 1.7.1 instead of 1.7.0 when it becomes available (sorry!)

**Comment by rommsen** [comment]
Awesome. Will try later

**Comment by rommsen** [comment]
Awesome. Love it. Will probably use this API for pretty much all non queries. So flexible and composable

```F#
module PersistentReadmodels =
  open Infrastructure
  open Domain
  open Npgsql.FSharp

  let parameters eventEnvelope =
    match eventEnvelope.Event with
    | Flavour_sold (Truck truck,flavour) ->
        [
          "@truck", SqlValue.Uuid truck
          "@flavour", SqlValue.String (Flavour.toString flavour)
        ] |> Some

    | _ -> None

  let flavourSoldListener (DB_Connection_String db_connection) : EventListener<Event> =
    let query = """
      INSERT INTO flavours_sold (truck, flavour, sold) VALUES (@truck, @flavour, 1)
      ON CONFLICT (truck,flavour) DO UPDATE SET sold = flavours_sold.sold + 1"""

    fun eventEnvelopes ->
      let parameters =
        eventEnvelopes |> List.choose parameters

      if not <| List.isEmpty parameters then
        db_connection
        |> Sql.connect
        |> Sql.executeTransactionAsync [ query, parameters ]
        |> Async.Ignore
      else
        async { return () }
```

**Comment by Zaid-Ajaj** [comment]
Glad to hear this Roman! 

Though I just realized that there is a little bug in there: a query such as `DELETE FROM flavours_sold` will not execute if the parameters are empty:
```fs
for (query, parameterSets) in queries do
      for parameterSet in parameterSets do
          use command = new NpgsqlCommand(query, connection, transaction)
          populateRow command parameterSet
          let affectedRows = command.ExecuteNonQuery() 
          affectedRowsByQuery.Add affectedRows
```
Instead it should be:
```fs
for (query, parameterSets) in queries do
    if List.isEmpty parameterSets 
    then 
       use command = new NpgsqlCommand(query, connection, transaction)
       let affectedRows = command.ExecuteNonQuery() 
       affectedRowsByQuery.Add affectedRows
    else
      for parameterSet in parameterSets do
        use command = new NpgsqlCommand(query, connection, transaction)
        populateRow command parameterSet
        let affectedRows = command.ExecuteNonQuery() 
        affectedRowsByQuery.Add affectedRows
```
Just updated and published the fix as of version 1.9.0, please update :)

**Comment by Zaid-Ajaj** [comment]
I guess this is resolved, if any problems come up, let me know :smile:
